File: Advanced/Arrays/Logic Building/1_move_zeroes_to_end.py
# Move Zeros to End
# Brute Force Approach
# TC = O(n)
# SC = O(n)
def move_zeroes_to_end_brute(arr):
    n = len(arr)
    temp = []
    for i in range(n):
        if arr[i] != 0:
            temp.append(arr[i])
    for i in range(len(temp)):
        arr[i] = temp[i]
    for i in range(len(temp), n):
        arr[i] = 0


# Optimal Approach
# TC = O(n)
# SC = O(1)


# Version 1 (simpler, may self-swap)
def move_zeroes_to_end_optimal_v1(arr):
    n = len(arr)
    j = 0
    for i in range(n):
        if arr[i] != 0:
            arr[i], arr[j] = arr[j], arr[i]
            j += 1


# Version 2 (first-zero search, avoids self-swaps)
def move_zeroes_to_end_optimal_v2(arr):
    n = len(arr)
    j = -1
    for i in range(n):
        if arr[i] == 0:
            j = i
            break
    if j == -1:
        return
    for i in range(j + 1, n):
        if arr[i] != 0:
            arr[i], arr[j] = arr[j], arr[i]
            j += 1

File: Advanced/Arrays/Logic Building/1_move_zeroes_to_end.cpp
#include <bits/stdc++.h>
using namespace std;

// --- Brute Force Approach ---
// TC = O(n)
// SC = O(n)
void moveZeroesToEndBrute(vector<int> &arr)
{
    int n = arr.size();
    vector<int> temp;

    for (int i = 0; i < n; i++)
    {
        if (arr[i] != 0)
        {
            temp.push_back(arr[i]);
        }
    }

    for (int i = 0; i < (int)temp.size(); i++)
    {
        arr[i] = temp[i];
    }

    for (int i = temp.size(); i < n; i++)
    {
        arr[i] = 0;
    }
}

// --- Optimal Approach (Version 1) ---
// TC = O(n)
// SC = O(1)
// Simpler, may self-swap
void moveZeroesToEndOptimalV1(vector<int> &arr)
{
    int n = arr.size();
    int j = 0;
    for (int i = 0; i < n; i++)
    {
        if (arr[i] != 0)
        {
            swap(arr[i], arr[j]);
            j++;
        }
    }
}

// --- Optimal Approach (Version 2) ---
// TC = O(n)
// SC = O(1)
// Avoids self-swaps by finding first zero index
void moveZeroesToEndOptimalV2(vector<int> &arr)
{
    int n = arr.size();
    int j = -1;

    // Find index of first zero
    for (int i = 0; i < n; i++)
    {
        if (arr[i] == 0)
        {
            j = i;
            break;
        }
    }

    if (j == -1)
        return; // No zeros found

    for (int i = j + 1; i < n; i++)
    {
        if (arr[i] != 0)
        {
            swap(arr[i], arr[j]);
            j++;
        }
    }
}

// --- Helper to Print Array ---
void printArray(const vector<int> &arr)
{
    for (int num : arr)
        cout << num << " ";
    cout << endl;
}

// --- Main Function ---
int main()
{
    vector<int> arr1 = {0, 1, 0, 3, 12};
    vector<int> arr2 = arr1;
    vector<int> arr3 = arr1;

    cout << "Original Array: ";
    printArray(arr1);

    moveZeroesToEndBrute(arr1);
    cout << "After Brute Force: ";
    printArray(arr1);

    moveZeroesToEndOptimalV1(arr2);
    cout << "After Optimal V1: ";
    printArray(arr2);

    moveZeroesToEndOptimalV2(arr3);
    cout << "After Optimal V2: ";
    printArray(arr3);

    return 0;
}
